primitives/beefy/src/mmr.rs

```diff
@@ -26,9 +26,23 @@
 //! but we imagine they will be useful for other chains that either want to bridge with Polkadot
 //! or are completely standalone, but heavily inspired by Polkadot.
 
+use crate::Vec;
 use codec::{Decode, Encode};
 use scale_info::TypeInfo;
 
+/// A provider for extra data that gets added to the Mmr leaf
+pub trait BeefyDataProvider {
+	/// Return a vector of bytes, ideally should be a merkle root hash
+	fn extra_data() -> Vec<u8>;
+}
+
+/// A default implementation for runtimes.
+impl BeefyDataProvider for () {
+	fn extra_data() -> Vec<u8> {
+		Vec::new()
+	}
+}
+
 /// A standard leaf that gets added every block to the MMR constructed by Substrate's `pallet_mmr`.
 #[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
 pub struct MmrLeaf<BlockNumber, Hash, MerkleRoot> {
```

Comment:
@acatangiu Do you think we still need `version` field here? I mean - now everything that may be changed, can be isolated within this `leaf_extra` field, right? So probably versions must be moved there?

Reply:
It's my understanding that this `version` is for the whole `MmrLeaf` struct, and it's there to allow us adding new fields potentially needed for generic functionality.
While `leaf_extra` is _custom_ data where downstream pallets can encode their non-generic functionality (probably the data encoded in `leaf_extra` should also encode a version for itself).

So unless we know that `parent_number_and_hash` and `beefy_next_authority_set` are all we'll ever need for the generic BEEFY functionality (I'm not confident enough yet to say that for a fact), I would keep the `version`.